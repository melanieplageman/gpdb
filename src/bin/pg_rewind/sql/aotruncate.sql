#!/bin/bash

# This file has the .sql extension, but it is actually launched as a shell
# script. This contortion is necessary because pg_regress normally uses
# psql to run the input scripts, and requires them to have the .sql
# extension, but we use a custom launcher script that runs the scripts using
# a shell instead.

TESTNAME=aotruncate

. sql/config_test.sh

# Nothing to do here
function before_master
{
:
}

# Do an insert in master.
function before_standby
{
PGOPTIONS=${PGOPTIONS_UTILITY} $MASTER_PSQL <<EOF
CREATE TABLE aotable (d text) with (appendonly=true);
INSERT INTO aotable VALUES ('in master');
CHECKPOINT;
EOF
}

function standby_following_master
{
# Insert additional data on master that will be replicated to standby
PGOPTIONS=${PGOPTIONS_UTILITY} $MASTER_PSQL -c "INSERT INTO aotable values ('in master, before promotion');"

# Launch checkpoint after standby has been started
PGOPTIONS=${PGOPTIONS_UTILITY} $MASTER_PSQL -c "CHECKPOINT;"
}

# This script runs after the standby has been promoted. Old Master is still
# running.
function after_promotion
{
# Insert a row in the old master. This causes the master and standby to have
# "diverged", it's no longer possible to just apply the standy's logs over
# master directory - you need to rewind.
PGOPTIONS=${PGOPTIONS_UTILITY} $MASTER_PSQL -c "DELETE from aotable;"
PGOPTIONS=${PGOPTIONS_UTILITY} $MASTER_PSQL -c "VACUUM full aotable;"
PGOPTIONS=${PGOPTIONS_UTILITY} $MASTER_PSQL -c "SELECT * from aotable;"

# Also insert a new row in the standby, which won't be present in the old
# master.
PGOPTIONS=${PGOPTIONS_UTILITY} $STANDBY_PSQL -c "INSERT INTO aotable VALUES ('in standby, after promotion');"
}

# Compare results generated by querying new master after rewind
function after_rewind
{
PGOPTIONS=${PGOPTIONS_UTILITY} $STANDBY_PSQL -c "SELECT * from aotable;"
}

# Compare results generated by querying original master after rewind
# Specifically, the tuples which would have been eliminated by deleting from aotable have been restored
function after_master_promotion
{
PGOPTIONS=${PGOPTIONS_UTILITY} $MASTER_PSQL -c "SELECT * from aotable;"
}

# Run the test
. sql/run_test.sh
