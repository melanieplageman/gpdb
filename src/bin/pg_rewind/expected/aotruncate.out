Master initialized.
Master running.
CREATE TABLE aotable (d text) with (appendonly=true);
CREATE TABLE
INSERT INTO aotable VALUES ('in master');
INSERT 0 1
CHECKPOINT;
CHECKPOINT
Standby initialized and running.
INSERT INTO aotable values ('in master, before promotion');
INSERT 0 1
CHECKPOINT;
CHECKPOINT
Standby promoted.
DELETE from aotable;
DELETE 2
VACUUM full aotable;
VACUUM
SELECT * from aotable;
 d 
---
(0 rows)

INSERT INTO aotable VALUES ('in standby, after promotion');
INSERT 0 1
Running pg_rewind...
Old master restarted after rewind.
SELECT * from aotable;
              d              
-----------------------------
 in master
 in master, before promotion
 in standby, after promotion
(3 rows)

Master promoted.
SELECT * from aotable;
              d              
-----------------------------
 in master
 in master, before promotion
 in standby, after promotion
(3 rows)

